### ISSUE TITLE
Database migrations/restore failing, Uuid referenced not found in database

### ISSUE THREAD

**Issue by 0xC0ncord** [kanidm/kanidm#1594]
### I did this
I upgraded `kanidm-server` to `1.1.0-alpha12` and attempted to start it, but this resulted in the following errors at startup:
```
May 03 15:58:42 petra.fuwafuwatime.moe kanidmd[626988]: 015add86-a8d2-4473-9f23-84adcb46b3ad ERROR       ┝━ 🚨 [error]: UUID reference set size differs from query result size <fast path, no uuid info available> | event_tag_id: 1
May 03 15:58:42 petra.fuwafuwatime.moe kanidmd[626988]: 015add86-a8d2-4473-9f23-84adcb46b3ad ERROR       ┝━ 🚨 [error]: Post-Modify operation failed (plugin), Plugin(ReferentialIntegrity("Uuid referenced not found in database")) | event_tag_id: 1
May 03 15:58:42 petra.fuwafuwatime.moe kanidmd[626988]: 015add86-a8d2-4473-9f23-84adcb46b3ad ERROR       ┕━ 🚨 [error]: initialise_idm p3 -> result | event_tag_id: 1 | res: Err(Plugin(ReferentialIntegrity("Uuid referenced not found in database")))
May 03 15:58:42 petra.fuwafuwatime.moe kanidmd[626988]: 00000000-0000-0000-0000-000000000000 ERROR    🚨 [error]: Unable to setup query server or idm server -> Plugin(ReferentialIntegrity("Uuid referenced not found in database"))
May 03 15:58:42 petra.fuwafuwatime.moe kanidmd[626988]: 00000000-0000-0000-0000-000000000000 ERROR    🚨 [error]: Failed to start server core!
```
Subsequent attempts to start `kanidm-server` after these errors resulted in these errors:
```
May 03 15:59:09 petra.fuwafuwatime.moe kanidmd[627235]: ddf4ebc5-e705-40d4-8048-3fa441f71e27 WARN     ┝━ 🚧 [warn]: starting 9 to 10 migration. | event_tag_id: 2
May 03 15:59:09 petra.fuwafuwatime.moe kanidmd[627235]: ddf4ebc5-e705-40d4-8048-3fa441f71e27 INFO     ┝━ ｉ [info]: allowed 5 entries ✅ | event_tag_id: 10
May 03 15:59:09 petra.fuwafuwatime.moe kanidmd[627235]: ddf4ebc5-e705-40d4-8048-3fa441f71e27 INFO     ┝━ ｉ [info]: migrate_10_to_11 no entries to migrate, complete | event_tag_id: 3
May 03 15:59:09 petra.fuwafuwatime.moe kanidmd[627235]: ddf4ebc5-e705-40d4-8048-3fa441f71e27 WARN     ┝━ 🚧 [warn]: starting 11 to 12 migration. | event_tag_id: 2
May 03 15:59:09 petra.fuwafuwatime.moe kanidmd[627235]: ddf4ebc5-e705-40d4-8048-3fa441f71e27 INFO     ┝━ ｉ [info]: allowed 1 entries ✅ | event_tag_id: 10
May 03 15:59:09 petra.fuwafuwatime.moe kanidmd[627235]: ddf4ebc5-e705-40d4-8048-3fa441f71e27 ERROR    ┕━ 🚨 [error]: Failed to convert api_token_session from session -> apitoken
May 03 15:59:09 petra.fuwafuwatime.moe kanidmd[627235]: 00000000-0000-0000-0000-000000000000 ERROR    🚨 [error]: Unable to setup query server or idm server -> InvalidValueState
May 03 15:59:09 petra.fuwafuwatime.moe kanidmd[627235]: 00000000-0000-0000-0000-000000000000 ERROR    🚨 [error]: Failed to start server core!
```
It turns out that all of my database backups going back about 4 weeks are producing the same first batch of errors when trying to restore the database, even on version `1.1.0-alpha11`, so this isn't related to a problematic upgrade.

### I expected the following
Database migrations should have completed successfully and all of the database backups that are currently causing errors should be restoreable without problems.

### Kanidm version (and git commit)
I can reproduce this on both `1.1.0-alpha11` (d3a2a6b) and `1.1.0-alpha12` (bcdbb18).

### Operating System / Version
Gentoo Linux

```
Linux petra.fuwafuwatime.moe 6.1.20-gentoo-hardened1 #1 SMP Fri Mar 24 08:05:53 EDT 2023 x86_64 Intel(R) Xeon(R) CPU E5-2650L v4 @ 1.70GHz GenuineIntel GNU/Linux
```
### Any other comments
N/A

**Comment by 0xC0ncord** [comment]
> Have you tried a `kanidmd database reindex` or `verify`?

Yes, I did try these but I forgot to mention that I did with the errors, sorry.

`kanidmd database reindex -c /etc/kanidm/server.toml` results in:
```
e671dd57-4e50-4d98-ae05-d471a7f908a9 INFO     ┝━ ｉ [info]: allowed 1 entries ✅ | event_tag_id: 10
e671dd57-4e50-4d98-ae05-d471a7f908a9 WARN     ┝━ 🚧 [warn]: starting 9 to 10 migration. | event_tag_id: 2
e671dd57-4e50-4d98-ae05-d471a7f908a9 INFO     ┝━ ｉ [info]: allowed 5 entries ✅ | event_tag_id: 10
e671dd57-4e50-4d98-ae05-d471a7f908a9 INFO     ┝━ ｉ [info]: migrate_10_to_11 no entries to migrate, complete | event_tag_id: 3
e671dd57-4e50-4d98-ae05-d471a7f908a9 WARN     ┝━ 🚧 [warn]: starting 11 to 12 migration. | event_tag_id: 2
e671dd57-4e50-4d98-ae05-d471a7f908a9 INFO     ┝━ ｉ [info]: allowed 1 entries ✅ | event_tag_id: 10
e671dd57-4e50-4d98-ae05-d471a7f908a9 ERROR    ┕━ 🚨 [error]: Failed to convert api_token_session from session -> apitoken
00000000-0000-0000-0000-000000000000 ERROR    🚨 [error]: Unable to setup query server or idm server -> InvalidValueState
```
And `kanidmd database verify -c /etc/kanidm/server.toml` results in:
```
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: This is running as uid == 0 (root) which may be a security risk.               
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: WARNING: DB folder permissions on /var/lib/kanidm indicate it may not be RW. This could cause the server start up to fail!
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: Running in db verification mode ...                                                 
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: DB tickets -> 56                                                                                      
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: Profile -> release_suse_generic          
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: CPU Flags -> x86_64_v2                                                                                
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 243 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute domain_display_name was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute domain_name was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute domain_uuid was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute es256_private_key_der was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute fernet_private_key_str was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute badlist_password was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute displayname was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute jws_es256_private_key was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute displayname was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute jws_es256_private_key was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute primary_credential was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute user_auth_token_session was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute displayname was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute jws_es256_private_key was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute primary_credential was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute user_auth_token_session was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute dyngroup_filter was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute dyngroup_filter was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute grant_ui_hint was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute displayname was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute gidnumber was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute loginshell was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute mail was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute oauth2_consent_scope_map was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute oauth2_session was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute passkeys was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute primary_credential was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute unix_password was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute user_auth_token_session was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute displayname was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute es256_private_key_der was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute oauth2_allow_insecure_client_disable_pkce was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute oauth2_rs_basic_secret was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute oauth2_rs_name was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute oauth2_rs_origin was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute oauth2_rs_scope_map was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute oauth2_rs_token_key was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute displayname was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute es256_private_key_der was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute oauth2_allow_insecure_client_disable_pkce was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute oauth2_rs_basic_secret was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute oauth2_rs_name was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute oauth2_rs_origin was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute oauth2_rs_scope_map was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute oauth2_rs_token_key was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute api_token_session was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute displayname was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute jws_es256_private_key was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute primary_credential was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute gidnumber was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute displayname was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute es256_private_key_der was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute oauth2_allow_insecure_client_disable_pkce was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute oauth2_rs_basic_secret was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute oauth2_rs_name was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute oauth2_rs_origin was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute oauth2_rs_origin_landing was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute oauth2_rs_scope_map was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute oauth2_rs_token_key was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute displayname was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute es256_private_key_der was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute oauth2_rs_basic_secret was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute oauth2_rs_name was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute oauth2_rs_origin was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute oauth2_rs_scope_map was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute oauth2_rs_token_key was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute credential_update_intent_token was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute displayname was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute gidnumber was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute oauth2_consent_scope_map was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute oauth2_session was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute primary_credential was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute unix_password was not found in schema during replication request
00000000-0000-0000-0000-000000000000 WARN     🚧 [warn]: Attribute user_auth_token_session was not found in schema during replication request
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 243 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 243 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 247 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 243 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 1 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 4 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 2 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 8 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 11 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 2 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 1 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 2 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 2 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 1 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 1 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 1 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 2 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 1 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 2 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 2 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 1 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 2 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 1 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 1 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 2 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 2 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 1 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 1 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 1 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 2 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 1 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 1 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 1 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 1 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 1 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 1 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 1 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 1 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 10 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 1 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 4 entries ✅ | event_tag_id: 10
00000000-0000-0000-0000-000000000000 INFO     ｉ [info]: allowed 66 entries ✅ | event_tag_id: 10
Verification passed!
```

**Comment by yaleman** [comment]
Thanks for that, this looks like one of those internal things best handled by @Firstyear 😄

**Comment by Firstyear** [comment]
Sorry for the delay @0xC0ncord - I was traveling for work and feeling a bit unwell.

So to understand correctly. 1.1.0-alpha11 when given a backup that was created from 1.1.0-alpha11 will cause this error?

Can you attach the output of "RUST_LOG=trace" when performing the restore? There MAY be private info in that output - feel free to email me the output directly to william at blackhats.net.au for me to review (and then destroy).

**Comment by Firstyear** [comment]
@0xC0ncord Are you also willing/able to test from a git clone / cargo build? That will enable some extra options for us to pursue.

**Comment by Firstyear** [comment]
Okay, I have investigated the logs that were sent to me by @0xC0ncord and also have read through some of the code paths and have worked out there are two issues occurring here that seem to be coming together in some manner.

I'll start with the restoring of backups from .11. This was erroring because if the backup that had been taken already had performed the session to api token migration, it was being *re-applied*. I am not 100% sure why this is since the application of this migration *should* result in a server version bump to prevent double execution of this. To harden this path, I have added a dummy version of the migrate call into the api session so that duplicate calls results in an Ok rather than an Err which will allow the restore to proceed.

The second is the reason the upgrade failed to start with which is the referential integrity issue. What is occurring is during the upgrade when the idm all persons group is touched it triggers a post modification plugin handler. This then executes in order referential integrity, dyn group, and then memberof. The reason they are in this order is that memberof softly relies on references being valid, so we do refint first before mo. dyngroups and mo are bound together as when a dynamic group changes, we have to also retrigger memberof. 

During the upgrade refint is triggered and the member attribute has 4 uuids, and it appears that one or more of these no longer exists in the database. I'd be curious for you to check in your backup json the existance of the following 4 uuids on persons or accounts:

```
14f4a2c1-4541-4018-9e7d-9b81010ece6c
97050b26-a4e4-4d84-bcb6-eb47903c7f6d
a1ffa32e-5206-4203-b501-003b427be2ed
d3e9d011-be7d-4633-a4a1-2879106075eb
```

My current suspect is that one or more of these accounts has been removed or is missing, so the member attribute now becomes invalid since the uuid of that missing entry was NOT removed during the post delete phase.

The question now becomes how to proceed:

@0xC0ncord for you, to immediately proceed, I will ask you to try a patch ontop of .11 *or* .12, whichever you prefer that will allow you to restore from the backup. In addition, please check the 4 uuid's mentioned here if they exist. In your backup json, please remove the member attribute from `idm_all_persons` as this attribute is dynamically recalculated so no data will be lost and this will allow you to proceed. 

For us in Kanidm, we need to consider how to handle this. We don't want people to have to resort to editing their backups to fix this. 

Either we can have it so that when we restore from a backup, that entries that contain a dyngroup have their member attribute purged. 

Alternately, we can semi-harden refint for all code paths by having an exception for member attribute on dyngroups to prevent checking of their validity since member will always be recalculated on a modify path of one of these entries.

I think the latter is a better course of action since it will prevent the issue from appearing at all and means we self-heal at runtime without errors for the user. 

I am extremely puzzled by how you got into this situation in the first place, but right now the best we can do is have better self-healing to un-stick others who hit this (apparently rare) situation.

**Comment by Firstyear** [comment]
Thinking about it there is also an argument to ignore memberOf on refint code paths too in these cases in case of error where a uuid is added to memberOf or an entry was deleted somehow that is restored. I think this also would be valid in our self-healing philosophy since mo is also regenerated subsequently in this case.

**Comment by Firstyear** [comment]
Okay, the two PR's are there, once reviewed I'll make a branch for you to test against .12 if that's okay @0xC0ncord

**Comment by Firstyear** [comment]
@0xC0ncord The backport to .12 will be done tomorrow by @yaleman as I'm currently traveling. If you have any issues after that we can re-open this. Also I want to apologise that you hit this issue in the first place - we try to ensure that upgrades are always painless and correct and I'm sorry it didn't happen for you.

**Comment by yaleman** [comment]
Reopening until I push the backport 😄

**Comment by 0xC0ncord** [comment]
@Firstyear @yaleman I'm happy to report that after making the manual edit to the backup and applying those 2 PRs on top of 1.1.0-alpha12, the database restore was successful and `kanidmd` was able to start. :)

> I'd be curious for you to check in your backup json the existance of the following 4 uuids on persons or accounts:

It seems that UUID `a1ffa32e-5206-4203-b501-003b427be2ed` in fact does not exist in the backup - I only see that UUID being referenced by `idm_all_persons` and `idm_all_accounts` in the `member` attribute.

> Also I want to apologise that you hit this issue in the first place - we try to ensure that upgrades are always painless and correct and I'm sorry it didn't happen for you.

No worries! Thankfully this was a relatively small deployment just in my own home lab where I'm currently trying to enable SSO for all my applications. I'm glad I was able to help identify the issue and get it fixed.

Thank you both! :)

**Comment by yaleman** [comment]
The fixes have been backported to 1.1.0-alpha.12  now. 🥳

**Comment by Firstyear** [comment]
@0xC0ncord If you have any other issues or feedback at all, please let us know!
